core/renderers/zelos/measurables/rows.js

```diff
@@ -0,0 +1,131 @@
+/**
+ * @license
+ * Visual Blocks Editor
+ *
+ * Copyright 2019 Google Inc.
+ * https://developers.google.com/blockly/
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview An object representing a single row on a rendered block and all
+ * of its subcomponents.
+ * @author samelh@google.com (Sam El-Husseini)
+ */
+'use strict';
+
+goog.provide('Blockly.zelos.BottomRow');
+goog.provide('Blockly.zelos.TopRow');
+
+goog.require('Blockly.blockRendering.BottomRow');
+goog.require('Blockly.blockRendering.TopRow');
+
+
+/**
+ * An object containing information about what elements are in the top row of a
+ * block as well as sizing information for the top row.
+ * Elements in a top row can consist of corners, hats, spacers, and previous
+ * connections.
+ * After this constructor is called, the row will contain all non-spacer
+ * elements it needs.
+ * @package
+ * @constructor
+ * @extends {Blockly.blockRendering.TopRow}
+ */
+Blockly.zelos.TopRow = function() {
+  Blockly.zelos.TopRow.superClass_.constructor.call(this);
+};
+goog.inherits(Blockly.zelos.TopRow, Blockly.blockRendering.TopRow);
+
+/**
+ * Create all non-spacer elements that belong on the top row.
+ * @param {!Blockly.BlockSvg} block The block whose top row this represents.
+ * @package
+ * @override
+ */
+Blockly.zelos.TopRow.prototype.populate = function(block) {
```

Comment:
Do you have opinions on whether this and the bottomRow `populate` function belong in info instead?

Reply:
Yeah I think that makes sense. Top and bottom rows are the only measurables that are also defining what elements go inside of them in the class definition. InputRows on the other hand rely on ``info.createRows_`` to define what goes inside of them.

Reply:
https://github.com/google/blockly/issues/2945

You can do it now or I can do it as a followup once this lands.